clippy_lints/src/methods/method_without_self_relation.rs

```diff
@@ -0,0 +1,116 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_hir::{FnSig, ImplItem, ImplItemKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty};
+use rustc_span::Span;
+
+use super::METHOD_WITHOUT_SELF_RELATION;
+
+/// Check if a type contains a reference to `Self` anywhere in its structure.
+/// This includes direct references and generic parameters.
+fn contains_self<'tcx>(ty: Ty<'tcx>, self_ty: Ty<'tcx>) -> bool {
```

Comment:
Could this use `Ty::walk` to simplify the match here since that gives you an iterator over all types contained, or does this intentionally skip certain types in the `match`?